src/sentry/eventstream/kafka/backend.py

```diff
@@ -258,88 +300,89 @@ def handle_shutdown_request(signum: int, frame: Any) -> None:
+                error = message.error()
+                if error is not None:
+                    raise Exception(error)
 
-            i = i + 1
-            owned_partition_offsets[key] = message.offset() + 1
+                key = (message.topic(), message.partition())
+                if key not in owned_partition_offsets:
+                    logger.warning("Skipping message for unowned partition: %r", key)
+                    continue
 
-            use_kafka_headers = options.get("post-process-forwarder:kafka-headers")
+                i = i + 1
+                owned_partition_offsets[key] = message.offset() + 1
+                process_message(message)
 
-            if use_kafka_headers is True:
-                try:
-                    with self.sampled_eventstream_timer(
-                        instance="get_task_kwargs_for_message_from_headers"
-                    ):
-                        task_kwargs = get_task_kwargs_for_message_from_headers(message.headers())
-
-                    if task_kwargs is not None:
-                        with self.sampled_eventstream_timer(
-                            instance="dispatch_post_process_group_task"
-                        ):
-                            if task_kwargs["group_id"] is None:
-                                metrics.incr(
-                                    "eventstream.messages",
-                                    tags={"partition": message.partition(), "type": "transactions"},
-                                )
-                            else:
-                                metrics.incr(
-                                    "eventstream.messages",
-                                    tags={"partition": message.partition(), "type": "errors"},
-                                )
-                            self._dispatch_post_process_group_task(**task_kwargs)
-
-                except Exception as error:
-                    logger.error("Could not forward message: %s", error, exc_info=True)
-                    self._get_task_kwargs_and_dispatch(message)
-
-            else:
-                self._get_task_kwargs_and_dispatch(message)
-
-            if i % commit_batch_size == 0:
-                commit_offsets()
+        # The main loop of the function. It will run until shutdown is requested.
+        # There is an option provided which can switch the number of threads used.
+        # We don't want to switch while running inside the implementation loop since
+        # that would increase code complexity.
+        while not shutdown_requested:
+            new_concurrency = options.get("post-process-forwarder:concurrency")
+            if new_concurrency != current_concurrency:
+                logger.info(f"Switching post-process-forwarder to {new_concurrency} worker threads")
+                executor.shutdown(wait=True)
+                executor = ThreadPoolExecutor(max_workers=new_concurrency)
+                current_concurrency = new_concurrency
```

Comment:
Curious why you chose to provide this via option instead of via CLI argument? Option makes sense to me if it is a very temporary thing that we want to test and later hardcode a value and remove the option. Otherwise I think CLI arg is better for something permanent.

Reply:
Good question. Ultimately, this would become a CLI option. But we want to be able to change it multiple times during stress test and be able to scale it down immediately if systems start behaving badly. Hence kept it as an option. I will open a task to convert it to CLI option in the future